### ISSUE TITLE
Basic design

### ISSUE THREAD

**Issue by ikeycode** [serpent-os/usysconf-rs#1]
Jotting down prior to a meeting..

Aim for Solus support first, with reusable core library for moss

 - `bincode` for the state DB
 - `clap` for CLI handling
 - `ion-rs` for trigger compilation into shared trigger state (cascade stateless)
 - `ion-schema` for CLI tool to validate triggers at *package build time*

Use a digraph and likely `rayon` to get best runtimes from usysconf. Less time doing repeat work.

Experimenting initially with `ion` for our trigger definition due to strong typing guarantees

**Comment by EbonJaeger** [comment]
Out of curiosity, what would be different in a Solus vs Serpent implementation? Is there anything keeping it from being completely agnostic between the two, so both can benefit from it immediately?

**Comment by ikeycode** [comment]
I'd like to share as *much* as possible. Main difference I think is that the Solus implementation needs the wrapper binary and has no path knowledge, as pisi can't pass that stuff along in terms of "what changed" in a trivial fashion.

moss would directly link usysconf crate and notify of the changes, ie not needing the mtime approach

**Comment by EbonJaeger** [comment]
Ah, okay. Yeah, that would be really nice for `moss`. Basically, the binary would take the place of `moss` in that environment.

**Comment by ikeycode** [comment]
Right, think of `/usr/sbin/usysconf` as a moss-adapter. Internally they'll both do the same thing and have 100% compatibility - pushing both projects forward. ^_^

**Comment by EbonJaeger** [comment]
Awesome! That makes perfect sense.

**Comment by livingsilver94** [comment]
> * ion-rs for trigger compilation into shared trigger state (cascade stateless)
> * ion-schema for CLI tool to validate triggers at package build time

We almost unanimously chose to use YAML + schema validation instead of Ion.

**Comment by chrisdebian** [comment]
Can anyone update this issue? A lot of (alpha release) work has taken place, and it would be useful to understand if it's still relevant.

Chris
